swarm/src/connection/handler_wrapper.rs

```diff
@@ -307,27 +307,57 @@ where
             Error<TProtoHandler::Error>,
         >,
     > {
-        while let Poll::Ready(Some((user_data, res))) = self.negotiating_in.poll_next_unpin(cx) {
-            match res {
-                Ok(upgrade) => self
-                    .handler
-                    .inject_fully_negotiated_inbound(upgrade, user_data),
-                Err(err) => self.handler.inject_listen_upgrade_error(user_data, err),
+        loop {
+            // Poll the [`ConnectionHandler`].
+            if let Poll::Ready(res) = self.handler.poll(cx) {
+                match res {
+                    ConnectionHandlerEvent::Custom(event) => {
+                        return Poll::Ready(Ok(Event::Custom(event)));
+                    }
+                    ConnectionHandlerEvent::OutboundSubstreamRequest { protocol } => {
+                        let id = self.unique_dial_upgrade_id;
+                        let timeout = *protocol.timeout();
+                        self.unique_dial_upgrade_id += 1;
+                        let (upgrade, info) = protocol.into_upgrade();
+                        self.queued_dial_upgrades.push((id, SendWrapper(upgrade)));
+                        return Poll::Ready(Ok(Event::OutboundSubstreamRequest((
+                            id, info, timeout,
+                        ))));
+                    }
+                    ConnectionHandlerEvent::Close(err) => return Poll::Ready(Err(err.into())),
+                }
             }
-        }
 
-        while let Poll::Ready(Some((user_data, res))) = self.negotiating_out.poll_next_unpin(cx) {
-            match res {
-                Ok(upgrade) => self
-                    .handler
-                    .inject_fully_negotiated_outbound(upgrade, user_data),
-                Err(err) => self.handler.inject_dial_upgrade_error(user_data, err),
+            // In case the [`ConnectionHandler`] can not make any more progress, poll the negotiating outbound streams.
+            if let Poll::Ready(Some((user_data, res))) = self.negotiating_out.poll_next_unpin(cx) {
+                match res {
+                    Ok(upgrade) => self
+                        .handler
+                        .inject_fully_negotiated_outbound(upgrade, user_data),
+                    Err(err) => self.handler.inject_dial_upgrade_error(user_data, err),
+                }
+
+                // After the `inject_*` calls, the [`ConnectionHandler`] might be able to make progress.
+                continue;
```

Comment:
I think we could also recurse here and thus avoid the loop? Not sure if that is necessarily easier to understand.

Reply:
I don't follow. What do you mean with "recurse"? As in a recursive call to the same method?

Reply:
Yeah. You should be able to just call `self.poll` here I believe?

Reply:
I have not seen recursion used on any serious Rust project. Not saying that it does not exist. Thus far I have avoided recursion in Rust due to:

- No Tail Call Ellimination
- Call stack limits
- Sensing that this is not idiomatic, thus not intuitive

Am I missing something? Do the above arguments not apply here?

Reply:
I think in this particular case, the recursion is very likely to stop after one call, is it not?

But I understand that this might be too much of a detail to notice and thus make the code harder to understand. Also, it is not guaranteed and thus risky to use with the usual problems of recursion.

Happy for the loop to stay, was just an idea :)